### DISCUSSION TITLE
Need help building with pre-installed PowerShell packages for Arch Linux

### DISCUSSION THREAD

**Discussion by carlsmedstad** [PowerShell/PowerShell#19311] [Q&A]
Hey :wave: 

I'm one of the maintainers for the Arch User Repository (AUR) package for PowerShell. Our current package does not include any of the PowerShell packages that come pre-installed in the released executable (ex. PowerShellGet and PSReadLine).

We need some help with figuring out to include these PowerShell packages when building from source. What step/s are we missing?

For reference our package is defined here: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=powershell

The key part is in the `build` function:

```sh
build() {
  cd "$srcdir/$_powershell_archive"

  export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=true
  export DOTNET_CLI_TELEMETRY_OPTOUT=true

  # Mock git describe output that the build expects
  export POWERSHELL_GIT_DESCRIBE_OUTPUT="v$pkgver-0-gxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"

  ## Restore
  dotnet restore src/powershell-unix
  dotnet restore src/ResGen
  dotnet restore src/TypeCatalogGen

  ## Setup the build target to gather dependency information
  cp "$srcdir/Microsoft.PowerShell.SDK.csproj.TypeCatalog.targets" "src/Microsoft.PowerShell.SDK/obj/Microsoft.PowerShell.SDK.csproj.TypeCatalog.targets"
  dotnet msbuild src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj /t:_GetDependencies "/property:DesignTimeBuild=true;_DependencyFile=$(pwd)/src/TypeCatalogGen/powershell.inc" /nologo

  ## create the telemetry flag file
  touch "$(pwd)/DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY"

  ## Generate resource binding C# files
  pushd src/ResGen
  dotnet run
  popd

  ## Generate 'CorePsTypeCatalog.cs'
  pushd src/TypeCatalogGen
  dotnet run ../System.Management.Automation/CoreCLR/CorePsTypeCatalog.cs powershell.inc
  popd

  ## Build native component
  pushd "$srcdir/$_powershell_native_archive/src/libpsl-native"
  cmake -DCMAKE_BUILD_TYPE=Debug .
  make -j
  popd

  ## Build powershell core
  dotnet publish --self-contained --configuration Linux "src/powershell-unix/" --output bin --runtime "linux-x64"
}
```

Explanation of the variables:

* `$_powershell_archive` is the location of a source archive of this repository.
* `$_powershell_native_archive` is the location of source archive of the repository https://github.com/PowerShell/PowerShell-Native.

My knowledge about .Net/PowerShell build processes is fairly limited so help is very much appreciated. Thanks!

**Comment by rhubarb-geek-nz** [comment]
This is just a thought...

As these are in the `Modules` directory they are technically addons so don't have to be part of the original build

```
/opt/microsoft/powershell/7/Modules $ ls
Microsoft.PowerShell.Archive     Microsoft.PowerShell.Security  PackageManagement
Microsoft.PowerShell.Host        Microsoft.PowerShell.Utility   PowerShellGet
Microsoft.PowerShell.Management  PSReadLine                     ThreadJob
```

I see no reason why they can't be compiled with a separate process and then copied into the directory. As far as I know these work like any other module on the `$Env:PSModulePath`

**Comment by rhubarb-geek-nz** [comment]
I notice that you hard-code the `linux-x64`... My cheap and cheerful approach uses the existing binary [PKGBUILD](https://sourceforge.net/p/rhubarb-pi/code/HEAD/tree/branches/pacman/pkg/powershell/PKGBUILD)
